README.md
(This is @viksit's port - non standard.)

# A reasonable Emacs config

This is my emacs configuration tree, continually used and tweaked
since 2000, and it may be a good starting point for other Emacs
users, especially those who are web developers. These days it's
somewhat geared towards OS X, but it is known to also work on Linux
and Windows.

Emacs itself comes with support for many programming languages. This
config adds improved defaults and extended support for the following:

* Ruby / Ruby on Rails
* CSS / LESS / SASS / SCSS
* HAML / Markdown / Textile / ERB
* Clojure (via nrepl)
* Javascript / Coffeescript
* Python
* PHP
* Haskell
* Erlang
* Common Lisp (with Slime)

In particular, there's a nice config for *tab autocompletion*, and
flycheck is used to immediately highlight syntax errors in Ruby, HAML,
Python, Javascript, PHP and a number of other languages.

## Requirements

* Emacs 23 or greater (note that Emacs 24 is required for some
  functionality, and will likely become the minimum required version
  some time soon.)

## Installation

To install, clone this repo to `~/.emacs.d`, i.e. ensure that the
`init.el` contained in this repo ends up at `~/.emacs.d/init.el`:

```
git clone https://github.com/purcell/emacs.d.git ~/.emacs.d
```

Upon starting up Emacs for the first time, further third-party
packages will be automatically downloaded and installed.

## Updates

Update the config with `git pull`. You'll probably also want/need to update
the third-party packages regularly too:

<kbd>M-x package-list-packages</kbd>, then <kbd>U</kbd> followed by <kbd>x</kbd>.

## Adding your own customization

To add your own customization, use <kbd>M-x customize</kbd> and/or
create a file `~/.emacs.d/lisp/init-local.el` which looks like this:

```el
... your code here ...

(provide 'init-local)
```

Alternatively, fork the repo and hack away at the config to make it your own!

## Similar configs

You might also want to check out `emacs-starter-kit` and `emacs-prelude`.

## Support / issues

If you hit any problems, please first ensure that you are using the latest version
of this code, and that you have updated your packages to the most recent available
versions (see "Updates" above). If you still experience problems, go ahead and
[file an issue on the github project](https://github.com/purcell/emacs.d).

-Steve Purcell

<hr>

[![](http://api.coderwall.com/purcell/endorsecount.png)](http://coderwall.com/purcell)

[![](http://www.linkedin.com/img/webpromo/btn_liprofile_blue_80x15.png)](http://uk.linkedin.com/in/stevepurcell)


eshell/history
uname -a
clear
ls
ssh ubuntu@gpu-t
ssh ubuntu@gpu-tf


color-theme/color-theme-autoloads.el
;;; -*-emacs-lisp-*-

;; Copyright (C) 2002 Free Software Foundation, Inc.

(defvar generated-autoload-file)
(defvar command-line-args-left)
(defun color-theme-generate-autoloads ()
  (interactive)
  (require 'autoload)
  (setq generated-autoload-file (car command-line-args-left))
  (setq command-line-args-left (cdr command-line-args-left))
  (batch-update-autoloads))

(provide 'color-theme-autoloads)
;;; Generated autoloads follow (made by autoload.el).

;;;### (autoloads nil nil ("themes/color-theme-example.el" "themes/color-theme-library.el")
;;;;;;  (19662 7608 626268))

;;;***

;;;### (autoloads (color-theme-initialize color-theme-submit color-theme-install
;;;;;;  color-theme-compare color-theme-make-snapshot color-theme-analyze-defun
;;;;;;  color-theme-print color-theme-install-at-point-for-current-frame
;;;;;;  color-theme-install-at-mouse color-theme-describe color-theme-select)
;;;;;;  "color-theme" "color-theme.el" (17529 41105))
;;; Generated autoloads from color-theme.el

(autoload (quote color-theme-select) "color-theme" "\
Displays a special buffer for selecting and installing a color theme.
With optional prefix ARG, this buffer will include color theme libraries
as well.  A color theme library is in itself not complete, it must be
used as part of another color theme to be useful.  Thus, color theme
libraries are mainly useful for color theme authors.

\(fn &optional ARG)" t nil)

(autoload (quote color-theme-describe) "color-theme" "\
Describe color theme listed at point.
This shows the documentation of the value of text-property color-theme
at point.  The text-property color-theme should be a color theme
function.  See `color-themes'.

\(fn)" t nil)

(autoload (quote color-theme-install-at-mouse) "color-theme" "\
Install color theme clicked upon using the mouse.
First argument EVENT is used to set point.  Then
`color-theme-install-at-point' is called.

\(fn EVENT)" t nil)

(autoload (quote color-theme-install-at-point-for-current-frame) "color-theme" "\
Install color theme at point for current frame only.
Binds `color-theme-is-global' to nil and calls
`color-theme-install-at-point'.

\(fn)" t nil)

(autoload (quote color-theme-print) "color-theme" "\
Print the current color theme function.

You can contribute this function to <URL:news:gnu.emacs.sources> or
paste it into your .emacs file and call it.  That should recreate all
the settings necessary for your color theme.

Example:

    (require 'color-theme)
    (defun my-color-theme ()
      \"Color theme by Alex Schroeder, created 2000-05-17.\"
      (interactive)
      (color-theme-install
       '(...
	 ...
	 ...)))
    (my-color-theme)

If you want to use a specific color theme function, you can call the
color theme function in your .emacs directly.

Example:

    (require 'color-theme)
    (color-theme-gnome2)

\(fn &optional BUF)" t nil)

(autoload (quote color-theme-analyze-defun) "color-theme" "\
Once you have a color-theme printed, check for missing faces.
This is used by maintainers who receive a color-theme submission
and want to make sure it follows the guidelines by the color-theme
author.

\(fn)" t nil)

(autoload (quote color-theme-make-snapshot) "color-theme" "\
Return the definition of the current color-theme.
The function returned will recreate the color-theme in use at the moment.

\(fn)" nil nil)

(autoload (quote color-theme-compare) "color-theme" "\
Compare two color themes.
This will print the differences between installing THEME-A and
installing THEME-B.  Note that the order is important: If a face is
defined in THEME-A and not in THEME-B, then this will not show up as a
difference, because there is no reset before installing THEME-B.  If a
face is defined in THEME-B and not in THEME-A, then this will show up as
a difference.

\(fn THEME-A THEME-B)" t nil)

(autoload (quote color-theme-install) "color-theme" "\
Install a color theme defined by frame parameters, variables and faces.

The theme is installed for all present and future frames; any missing
faces are created.  See `color-theme-install-faces'.

THEME is a color theme definition.  See below for more information.

If you want to install a color theme from your .emacs, use the output
generated by `color-theme-print'.  This produces color theme function
which you can copy to your .emacs.

A color theme definition is a list:
\([FUNCTION] FRAME-PARAMETERS VARIABLE-SETTINGS FACE-DEFINITIONS)

FUNCTION is the color theme function which called `color-theme-install'.
This is no longer used.  There was a time when this package supported
automatic factoring of color themes.  This has been abandoned.

FRAME-PARAMETERS is an alist of frame parameters.  These are installed
with `color-theme-install-frame-params'.  These are installed last such
that any changes to the default face can be changed by the frame
parameters.

VARIABLE-DEFINITIONS is an alist of variable settings.  These are
installed with `color-theme-install-variables'.

FACE-DEFINITIONS is an alist of face definitions.  These are installed
with `color-theme-install-faces'.

If `color-theme-is-cumulative' is nil, a color theme will undo face and
frame-parameter settings of previous color themes.

\(fn THEME)" nil nil)

(autoload (quote color-theme-submit) "color-theme" "\
Submit your color-theme to the maintainer.

\(fn)" t nil)

(autoload (quote color-theme-initialize) "color-theme" "\
Initialize the color theme package by loading color-theme-libraries.

\(fn)" t nil)

;;;***


init.el

;;; This file bootstraps the configuration, which is divided into
;;; a number of other files.

(let ((minver 23))
  (unless (>= emacs-major-version minver)
    (error "Your Emacs is too old -- this config requires v%s or higher" minver)))

(add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))
(require 'init-benchmarking) ;; Measure startup time

(defconst *spell-check-support-enabled* nil) ;; Enable with t if you prefer
(defconst *is-a-mac* (eq system-type 'darwin))

;;----------------------------------------------------------------------------
;; Bootstrap config
;;----------------------------------------------------------------------------
(require 'init-compat)
(require 'init-utils)
(require 'init-site-lisp) ;; Must come before elpa, as it may provide package.el
(require 'init-elpa)      ;; Machinery for installing required packages
(require 'init-exec-path) ;; Set up $PATH

;;----------------------------------------------------------------------------
;; Load configs for specific features and modes
;;----------------------------------------------------------------------------

(require-package 'wgrep)
(require-package 'project-local-variables)
(require-package 'diminish)
(require-package 'scratch)
(require-package 'mwe-log-commands)

;; (require 'init-frame-hooks)
;; (require 'init-xterm)
(require 'init-themes)
;; (require 'init-osx-keys)
(require 'init-gui-frames)
(require 'init-proxies)
(require 'init-dired)
(require 'init-isearch)
(require 'init-uniquify)
(require 'init-ibuffer)
(require 'init-flycheck)

(require 'init-recentf)
(require 'init-ido)
(require 'init-hippie-expand)
(require 'init-auto-complete)
(require 'init-windows)
(require 'init-sessions)
(require 'init-fonts)
(require 'init-mmm)

(require 'init-editing-utils)

;; (require 'init-darcs)
(require 'init-git)

(require 'init-crontab)
;; (require 'init-textile)
(require 'init-markdown)
(require 'init-csv)
;; (require 'init-erlang)
(require 'init-javascript)
(require 'init-php)
(require 'init-org)
;(require 'init-nxml)
(require 'init-css)
;(require 'init-haml)
; (require 'init-python-mode)
;(require 'init-haskell)
;(require 'init-ruby-mode)
;(require 'init-rails)
(require 'init-sql)

(require 'init-paredit)
(require 'init-lisp)
(require 'init-slime)
(require 'init-clojure)
(require 'init-common-lisp)

(when *spell-check-support-enabled*
  (require 'init-spelling))

(require 'init-marmalade)
(require 'init-misc)

(require 'init-dash)
(require 'init-ledger)
;; Extra packages which don't require any configuration

(require-package 'gnuplot)
;; (require-package 'lua-mode)
(require-package 'htmlize)
(require-package 'dsvn)
(when *is-a-mac*
  (require-package 'osx-location))
(require-package 'regex-tool)

;;----------------------------------------------------------------------------
;; Allow access from emacsclient
;;----------------------------------------------------------------------------
;; (require 'server)
;; (unless (server-running-p)
;;  (server-start))


;;----------------------------------------------------------------------------
;; Variables configured via the interactive 'customize' interface
;;----------------------------------------------------------------------------
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (file-exists-p custom-file)
  (load custom-file))


;;----------------------------------------------------------------------------
;; Allow users to provide an optional "init-local" containing personal settings
;;----------------------------------------------------------------------------
(when (file-exists-p (expand-file-name "init-local.el" user-emacs-directory))
  (error "Please move init-local.el to ~/.emacs.d/lisp"))
(require 'init-local nil t)


;;----------------------------------------------------------------------------
;; Locales (setting them earlier in this file doesn't work in X)
;;----------------------------------------------------------------------------
(require 'init-locales)

(add-hook 'after-init-hook
          (lambda ()
            (message "init completed in %.2fms"
                     (sanityinc/time-subtract-millis after-init-time before-init-time))))

;; ---- Tern
(add-hook 'js-mode-hook (lambda () (tern-mode t)))
(eval-after-load 'tern
   '(progn
      (require 'tern-auto-complete)
      (tern-ac-setup)))
;; -- End tern

;; (global-set-key [(control p)] 'simp-project-find-file)
(projectile-global-mode)
(setq projectile-enable-caching t)
(global-set-key [(control p)] 'projectile-find-file)

(add-hook 'before-save-hook 'delete-trailing-whitespace)

(setq ring-bell-function #'ignore)

;; (disable-paredit-mode)

;; (global-set-key "\C-c" 'comment-or-uncomment-region)

(global-set-key (kbd "C-x ;") 'comment-or-uncomment-region)
(global-set-key [C-tab] 'other-window)

;; ;; R shortcuts
;; ;; ESS Mode (.R file)
;; (define-key ess-mode-map "\C-e" 'ess-eval-line-and-step)
;; (define-key ess-mode-map "\C-p" 'ess-eval-function-or-paragraph-and-step)
;; (define-key ess-mode-map "\C-r" 'ess-eval-region)

;; ;; iESS Mode (R console)
;; (define-key inferior-ess-mode-map "\C-u" 'comint-kill-input)
;; (define-key inferior-ess-mode-map "\C-w" 'backward-kill-word)
;; (define-key inferior-ess-mode-map "\C-a" 'comint-bol)
;; (define-key inferior-ess-mode-map [home] 'comint-bol)

;; ;; Comint Mode (R console as well)
;; (define-key comint-mode-map "\C-e" 'comint-show-maximum-output)
;; (define-key comint-mode-map "\C-r" 'comint-show-output)
;; (define-key comint-mode-map "\C-o" 'comint-kill-output)
;; (setq-default indent-tabs-mode nil)
;; (setq-default tab-width 2)

;; (add-hook 'ess-mode-hook
;;  '(lambda()
;;        (setq indent-tabs-mode nil
;;                             tab-width 2)))

;; (when (eq window-system 'ns)
;;   (defadvice ns-get-pasteboard (around hack-empty-pasteboard compile activate)
;;     (condition-case err
;;         ad-do-it
;;       (quit (message "%s" (cadr err))
;;             nil))))

(setq cider-auto-select-error-buffer t)



;; Elpy mode for python
(elpy-enable)


;; (add-hook 'python-mode-hook
          ;; (function (lambda ()
          ;;             (setq indent-tabs-mode nil
          ;;                   tab-width 4))))


;;(add-hook 'python-mode-hook
;;          (lambda ()
;;            (setq indent-tabs-mode nil)
;;            (setq tab-width 4)
;;            (setq python-indent 4)))


;; (add-hook 'python-mode-hook
;;       (lambda ()
;;         (setq-default indent-tabs-mode nil)
;;         (setq-default tab-width 4)
;;         (setq-default python-indent 4)))
;; (setq-default indent-tabs-mode nil)
;; (setq-default tab-width 2)
;; (require 'python)
;; (setq python-shell-interpreter "ipython")
;; (setq python-shell-interpreter-args "--pylab")

(add-hook 'before-save-hook 'delete-trailing-whitespace)

(load-theme 'wombat t)

(define-globalized-minor-mode my-global-fixme-mode fixme-mode
  (lambda () (fixme-mode 1)))

(my-global-fixme-mode 1)

;; (require 'fic-mode)
;; emacs 24
;; (add-hook 'prog-mode-hook 'fic-mode)

;;(font-lock-add-keywords 'python-mode
;;  '(("pass" . font-lock-keyword-face)))

;; Local Variables:
;; coding: utf-8
;; no-byte-compile: t
;; End:

(provide 'init) ;; init.el ends here
